### PR TITLE
3D wavelet example

### PR DESCRIPTION
**Pull request by matthieutrs**
This PR adds a small 3D denoising example with wavelet denoising and showcases the implementation of Dystra-like for denoising in a dictionary of wavelets.

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: examples/optimization/demo_3D_wavelets.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+DATA_DIR = BASE_DIR / "measurements"
+RESULTS_DIR = BASE_DIR / "results"
+DEG_DIR = BASE_DIR / "degradations"
+
+
+# %%
+# Load base volume image and denoising operators.
+# ----------------------------------------------------------------------------------------
+# In this example, we use a T1-weighted brain MRI volume from the BrainWeb dataset (subject id 4) and we add
+# Gaussian random noise to it.
```
Comment (tachella) [inline examples/optimization/demo_3D_wavelets.py:36]: maybe explain the size of the tensor representing a volume?

is it B x C x H x W x P ?

File: examples/optimization/demo_3D_wavelets.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
+"""
+3D wavelet denoising
+====================================================================================================
+
+This example shows how to use a 3D wavelet denoiser for denoising a 3D image. We first apply a standard soft-thresholding
+wavelet denoiser to a 3D brain MRI volume. We then extend the denoiser objective to a redundant dictionary of wavelet
+bases, which does not admit a closed-form solution. We solve the denoising problem using the Dykstra-like algorithm.
```
Comment (tachella) [inline examples/optimization/demo_3D_wavelets.py:7]: The resulting algorithm is readily available in :meth:`deepinv.models.WaveletDictDenoiser` class

File: examples/optimization/demo_3D_wavelets.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+plt.subplot(133)
+plt.imshow(x_cur[0].squeeze()[..., 90])
+plt.suptitle("Denoised brain volume after 10 steps. PSNR = {:.2f}dB".format(psnr))
+plt.show()
+
+
+# %%
+# Using the Dykstra-like algorithm for wavelet denoising.
+# -------------------------------------------------------
+# You can readily use this algorithm via the :meth:`deepinv.models.WaveletDict` class.
```
Comment (tachella) [inline examples/optimization/demo_3D_wavelets.py:238]: WaveletDictDenoiser

**Review by tachella — CHANGES_REQUESTED** [review]
Great PR! I really like having 3D denoisers :-) 

I just flagged a few docs issues before merging

Also please see the comment on the plotting in the 3D example

File: deepinv/utils/demo.py [hunk 4]
```diff
@@ -169,3 +169,21 @@ def load_url_image(
     transform = transforms.Compose(transform_list)
     x = transform(img).unsqueeze(0).to(device)
     return x
+
+
+def load_url_data(
```
[the remark this answers is not in the document]
Comment (matthieutrs) [reply deepinv/utils/demo.py]: I've renamed to `load_np_url`, this is just for loading a np array

File: examples/optimization/demo_3D_wavelets.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+# of the wavelet prior and can be used as a denoiser. More precisely it solves the following problem
+#
+# .. math::
+#     \widehat{x} = \arg\min_{x} \frac{1}{2} \|y - x\|_2^2 + \lambda \|\Psi x\|_1 = \operatorname{prox}_{\lambda \|\Psi \cdot\|_1}(y)
+#
+# where :math:`\Psi` is the wavelet transform and :math:`\lambda` is the thresholding parameter. The solution to
+# this problem is given by the proximal operator of the wavelet prior.
+#
+# .. note::
+#     The computational complexity of the wavelet transform in 3D grows cubically with the size of the support. For this
```
[the remark this answers is not in the document]
Comment (matthieutrs) [reply examples/optimization/demo_3D_wavelets.py:87]: Yes, maybe indeed

File: deepinv/models/wavdict.py [hunk 6]
```diff
@@ -79,6 +79,28 @@ def flatten_coeffs(self, dec):
             )
         return flat
 
+    @staticmethod
+    def psi(x, wavelet="db2", level=2, dimension=2):
+        r"""
+        TODO: take only into account the detail coefficients for the loss computation
```
[the remark this answers is not in the document]
Comment (matthieutrs) [reply deepinv/models/wavdict.py]: Yes this is common practice to not threshold the approximation coefficients, but if you want to threshold them you can simply increase the level of decomposition

File: examples/optimization/demo_3D_wavelets.py (hunk 1 above)
Comment (matthieutrs) [reply examples/optimization/demo_3D_wavelets.py]: Done (the default torch convention is BxCxDxHxW (depth is after C)

**Review by tachella — APPROVED** [review]
great PR!

**Pull request merged**